File: Cookies.py
from datetime import datetime, timedelta, timezone
from Crypto.PublicKey import RSA
from Crypto.Signature import pkcs1_15
from Crypto.Hash import SHA1
import base64
import json

def rsa_signer(message, private_key_path):
    with open(private_key_path, 'rb') as f:
        private_key = RSA.import_key(f.read())
    h = SHA1.new(message.encode('utf-8'))
    signature = pkcs1_15.new(private_key).sign(h)
    return base64.b64encode(signature).decode('utf-8')

def create_signed_cookies(url_prefix, key_pair_id, private_key_path):
    expire_date = datetime.now(timezone.utc) + timedelta(hours=5)  # 1 hour expiration

    policy = {
        "Statement": [{
            "Resource": url_prefix,
            "Condition": {
                "DateLessThan": {"AWS:EpochTime": int(expire_date.timestamp())}
            }
        }]
    }

    policy_json = json.dumps(policy).replace(' ', '')
    signature = rsa_signer(policy_json, private_key_path)

    return {
        'CloudFront-Policy': base64.b64encode(policy_json.encode()).decode('utf-8'),
        'CloudFront-Signature': signature,
        'CloudFront-Key-Pair-Id': key_pair_id
    }

# Example usage
url_prefix = "https://d22si132od0d26.cloudfront.net/*"  # All content under distribution
key_pair_id = "KG3VA0PMQKICO"
private_key_path = "cookies.pem"

cookies = create_signed_cookies(url_prefix, key_pair_id, private_key_path)
for name, value in cookies.items():
    print(f"{name} = {value}")

File: cookies.js
const crypto = require('crypto');
const fs = require('fs').promises;
const path = require('path');

/**
 * Convert standard base64 to URL-safe base64
 */
function toUrlSafeBase64(str) {
  return Buffer.from(str)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

/**
 * Signs a message using RSA-SHA1 with the provided private key
 */
async function rsaSigner(message, privateKeyContent) {
  try {
    const sign = crypto.createSign('RSA-SHA1');
    sign.update(message);
    const signature = sign.sign(privateKeyContent);
    return signature.toString('base64');
  } catch (error) {
    throw new Error(`Failed to create signature: ${error.message}`);
  }
}

/**
 * Validates CloudFront config
 */
function validateConfig(config) {
  if (!config.urlPrefix || !config.urlPrefix.startsWith('https://')) {
    throw new Error('Invalid URL prefix. Must start with https://');
  }
  if (!config.keyPairId || config.keyPairId.length < 1) {
    throw new Error('Invalid key pair ID');
  }
  if (!config.privateKeyPath) {
    throw new Error('Private key path is required');
  }
  if (!config.expiryHours || config.expiryHours < 1) {
    throw new Error('Expiry hours must be greater than 0');
  }
}

/**
 * Creates signed cookies for CloudFront private content
 */
async function createSignedCookies(config) {
  try {
    validateConfig(config);

    const privateKeyContent = await fs.readFile(config.privateKeyPath);

    const expireDate = new Date();
    expireDate.setHours(expireDate.getHours() + config.expiryHours);

    const policy = {
      Statement: [
        {
          Resource: config.urlPrefix,
          Condition: {
            DateLessThan: {
              'AWS:EpochTime': Math.floor(expireDate.getTime() / 1000)
            }
          }
        }
      ]
    };

    const policyString = JSON.stringify(policy).replace(/\s/g, '');
    const encodedPolicy = toUrlSafeBase64(policyString);
    const rawSignature = await rsaSigner(policyString, privateKeyContent);
    const encodedSignature = toUrlSafeBase64(rawSignature);

    return {
      'CloudFront-Policy': encodedPolicy,
      'CloudFront-Signature': encodedSignature,
      'CloudFront-Key-Pair-Id': config.keyPairId
    };
  } catch (error) {
    throw new Error(`Failed to create signed cookies: ${error.message}`);
  }
}

module.exports = {
  createSignedCookies,
  validateConfig
};

